File: packages/go/contracts/domain/execution.go
package domain

import "time"

type OrderID string

type FillID string

type PositionID string

type AccountID string

type OrderSide string

type OrderType string

type OrderStatus string

type PositionDirection string

const (
	OrderSideBuy  OrderSide = "buy"
	OrderSideSell OrderSide = "sell"

	OrderTypeMarket    OrderType = "market"
	OrderTypeLimit     OrderType = "limit"
	OrderTypeStop      OrderType = "stop"
	OrderTypeStopLimit OrderType = "stop_limit"

	OrderStatusPending         OrderStatus = "pending"
	OrderStatusAccepted        OrderStatus = "accepted"
	OrderStatusPartiallyFilled OrderStatus = "partially_filled"
	OrderStatusFilled          OrderStatus = "filled"
	OrderStatusCanceled        OrderStatus = "canceled"
	OrderStatusRejected        OrderStatus = "rejected"

	PositionDirectionLong  PositionDirection = "long"
	PositionDirectionShort PositionDirection = "short"
	PositionDirectionFlat  PositionDirection = "flat"
)

type Fee struct {
	Amount   Decimal  `json:"amount"`
	Currency Currency `json:"currency"`
	Kind     string   `json:"kind,omitempty"`
}

type Order struct {
	ID             OrderID      `json:"id"`
	AccountID      AccountID    `json:"account_id,omitempty"`
	StrategyID     StrategyID   `json:"strategy_id,omitempty"`
	InstrumentID   InstrumentID `json:"instrument_id"`
	Venue          Venue        `json:"venue,omitempty"`
	Side           OrderSide    `json:"side"`
	Type           OrderType    `json:"type"`
	Status         OrderStatus  `json:"status"`
	Quantity       Decimal      `json:"quantity"`
	LimitPrice     Decimal      `json:"limit_price,omitempty"`
	StopPrice      Decimal      `json:"stop_price,omitempty"`
	FilledQuantity Decimal      `json:"filled_quantity,omitempty"`
	AverageFill    Decimal      `json:"average_fill,omitempty"`
	SubmittedAt    time.Time    `json:"submitted_at"`
	UpdatedAt      time.Time    `json:"updated_at"`
}

type Fill struct {
	ID           FillID       `json:"id"`
	OrderID      OrderID      `json:"order_id"`
	PositionID   PositionID   `json:"position_id,omitempty"`
	AccountID    AccountID    `json:"account_id,omitempty"`
	StrategyID   StrategyID   `json:"strategy_id,omitempty"`
	InstrumentID InstrumentID `json:"instrument_id"`
	Side         OrderSide    `json:"side"`
	Quantity     Decimal      `json:"quantity"`
	Price        Decimal      `json:"price"`
	Fees         []Fee        `json:"fees,omitempty"`
	ExecutedAt   time.Time    `json:"executed_at"`
}

type Position struct {
	ID            PositionID        `json:"id"`
	AccountID     AccountID         `json:"account_id,omitempty"`
	StrategyID    StrategyID        `json:"strategy_id,omitempty"`
	InstrumentID  InstrumentID      `json:"instrument_id"`
	Direction     PositionDirection `json:"direction"`
	Quantity      Decimal           `json:"quantity"`
	AverageOpen   Decimal           `json:"average_open_price,omitempty"`
	RealizedPNL   Decimal           `json:"realized_pnl,omitempty"`
	UnrealizedPNL Decimal           `json:"unrealized_pnl,omitempty"`
	OpenedAt      time.Time         `json:"opened_at"`
	ClosedAt      *time.Time        `json:"closed_at,omitempty"`
}

type Holding struct {
	InstrumentID  InstrumentID `json:"instrument_id"`
	Quantity      Decimal      `json:"quantity"`
	AveragePrice  Decimal      `json:"average_price,omitempty"`
	MarketPrice   Decimal      `json:"market_price,omitempty"`
	MarketValue   Decimal      `json:"market_value,omitempty"`
	CostBasis     Decimal      `json:"cost_basis,omitempty"`
	UnrealizedPNL Decimal      `json:"unrealized_pnl,omitempty"`
}

type PortfolioSnapshot struct {
	AccountID     AccountID `json:"account_id,omitempty"`
	Equity        Decimal   `json:"equity"`
	Cash          Decimal   `json:"cash"`
	BuyingPower   Decimal   `json:"buying_power,omitempty"`
	RealizedPNL   Decimal   `json:"realized_pnl,omitempty"`
	UnrealizedPNL Decimal   `json:"unrealized_pnl,omitempty"`
	Holdings      []Holding `json:"holdings,omitempty"`
	Timestamp     time.Time `json:"timestamp"`
}

File: packages/python/src/trend_buddy_shared/contracts/domain/execution.py
from dataclasses import dataclass, field
from datetime import datetime
from decimal import Decimal
from enum import StrEnum


class OrderSide(StrEnum):
    BUY = "buy"
    SELL = "sell"


class OrderType(StrEnum):
    MARKET = "market"
    LIMIT = "limit"
    STOP = "stop"
    STOP_LIMIT = "stop_limit"


class OrderStatus(StrEnum):
    PENDING = "pending"
    ACCEPTED = "accepted"
    PARTIALLY_FILLED = "partially_filled"
    FILLED = "filled"
    CANCELED = "canceled"
    REJECTED = "rejected"


class PositionDirection(StrEnum):
    LONG = "long"
    SHORT = "short"
    FLAT = "flat"


@dataclass
class Fee:
    amount: Decimal
    currency: str
    kind: str = ""


@dataclass
class Order:
    id: str
    instrument_id: str
    side: OrderSide
    type: OrderType
    status: OrderStatus
    quantity: Decimal
    submitted_at: datetime
    updated_at: datetime
    account_id: str = ""
    strategy_id: str = ""
    venue: str = ""
    limit_price: Decimal = Decimal("0")
    stop_price: Decimal = Decimal("0")
    filled_quantity: Decimal = Decimal("0")
    average_fill: Decimal = Decimal("0")


@dataclass
class Fill:
    id: str
    order_id: str
    instrument_id: str
    side: OrderSide
    quantity: Decimal
    price: Decimal
    executed_at: datetime
    position_id: str = ""
    account_id: str = ""
    strategy_id: str = ""
    fees: list[Fee] = field(default_factory=list)


@dataclass
class Position:
    id: str
    instrument_id: str
    direction: PositionDirection
    quantity: Decimal
    opened_at: datetime
    account_id: str = ""
    strategy_id: str = ""
    average_open: Decimal = Decimal("0")
    realized_pnl: Decimal = Decimal("0")
    unrealized_pnl: Decimal = Decimal("0")
    closed_at: datetime | None = None


@dataclass
class Holding:
    instrument_id: str
    quantity: Decimal
    average_price: Decimal = Decimal("0")
    market_price: Decimal = Decimal("0")
    market_value: Decimal = Decimal("0")
    cost_basis: Decimal = Decimal("0")
    unrealized_pnl: Decimal = Decimal("0")


@dataclass
class PortfolioSnapshot:
    equity: Decimal
    cash: Decimal
    timestamp: datetime
    account_id: str = ""
    buying_power: Decimal = Decimal("0")
    realized_pnl: Decimal = Decimal("0")
    unrealized_pnl: Decimal = Decimal("0")
    holdings: list[Holding] = field(default_factory=list)
